cmd/soroban-cli/src/config/self_outdated_check.rs

```diff
@@ -0,0 +1,82 @@
+use crate::config::locator;
+use jsonrpsee_core::Serialize;
+use semver::Version;
+use serde::Deserialize;
+use std::fs;
+
+const FILE_NAME: &str = "self_outdated_check.toml";
+
+/// The `SelfOutdatedCheck` struct represents the state of the self-outdated check.
+/// This state is global and stored in the `self_outdated_check.toml` file in
+/// the global configuration directory.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct SelfOutdatedCheck {
+    /// The timestamp of the latest check for a new version of the CLI.
+    pub latest_check_time: u64,
+    /// The latest stable version of the CLI available on crates.io.
+    pub max_stable_version: Version,
+    /// The latest version of the CLI available on crates.io, including pre-releases.
+    pub max_version: Version,
+}
+
+impl Default for SelfOutdatedCheck {
+    fn default() -> Self {
+        Self {
+            latest_check_time: 0,
+            max_stable_version: Version::new(0, 0, 0),
+            max_version: Version::new(0, 0, 0),
+        }
+    }
+}
+
+impl SelfOutdatedCheck {
+    /// Loads the state of the self-outdated check from the global configuration directory.
+    /// If the file doesn't exist, returns a default instance of `SelfOutdatedCheck`.
+    pub fn load() -> Result<Self, locator::Error> {
+        let path = locator::global_config_path()?.join(FILE_NAME);
+        if !path.exists() {
+            return Ok(Self::default());
+        }
+        let data = fs::read(&path)
+            .map_err(|error| locator::Error::SelfOutdatedCheckReadFailed { path, error })?;
+        Ok(toml::from_slice(data.as_slice())?)
+    }
+
+    /// Saves the state of the self-outdated check to the `self_outdated_check.toml` file in the global configuration directory.
+    pub fn save(&self) -> Result<(), locator::Error> {
+        let path = locator::global_config_path()?.join(FILE_NAME);
+        let path = locator::ensure_directory(path)?;
+        let data = toml::to_string(self).map_err(|_| locator::Error::ConfigSerialization)?;
```

Comment:
Can we use JSON for this file? The CLI uses JSON for most of the config files that get written, except for one, the key file that for some historical reason uses TOML.

Reply:
Why JSON over TOML?

Reply:
We should decide on a format, to maintain consistency in the project.

Reply:
We've used a mix of JSON and TOML for configs. Most of the "config" configs (internal, not in the user directory) are JSON. The update check config is another internal (not in the user directory).

In general, I think we should pick JSON unless there's value in picking something else. JSON is more accessible with a wide array of tooling without having to navigate the world of toml. Toml is popular, but the tooling isn't anywhere near as solid as JSON. (I had a case earlier today where a popular TOML lib couldn't render a JS Object, even though JSON could.)

There are some advantages to using TOML for very simple (single level) user facing config files, but this isn't a user facing config file.

Even for user facing config files, TOML can quickly become unintuitive for nested tables.

I'm not against toml in general, and they are easier on the eyes for simple flat config files. But JSON is sufficient, especially for internal config files.